std/protocols/permutation.asm

```diff
@@ -3,17 +3,76 @@ use std::array::fold;
 use std::array::len;
 use std::convert::int;
 use std::check::assert;
-use std::field::modulus;
+use std::check::panic;
+use std::field::known_field;
+use std::field::KnownField;
+use std::math::fp2::Fp2Value;
+use std::math::fp2::Fp2Expr;
+use std::math::fp2::add_ext;
+use std::math::fp2::sub_ext;
+use std::math::fp2::mul_ext;
+use std::math::fp2::unpack_ext;
+use std::math::fp2::next_ext;
+use std::math::fp2::eval_ext;
+use std::math::fp2::expr_ext;
+use std::math::fp2::inv_ext;
 
 let is_first: col = |i| if i == 0 { 1 } else { 0 };
 
 // Get two phase-2 challenges to use in all permutation arguments.
-// Note that this assumes that globally no other challenge of these IDs is used.
-let alpha: expr = challenge(0, 1);
-let beta: expr = challenge(0, 2);
+// Note that this assumes that globally no other challenge of these IDs is used,
+// and that challenges for multiple permutation arguments are re-used.
+// We declare two components for each challenge here, in case we need to operate
+// on the extension field. If we don't, we won't end up needing it and the optimizer
+// will remove it.
+let alpha1: expr = challenge(0, 1);
+let alpha2: expr = challenge(0, 2);
+
+let beta1: expr = challenge(0, 3);
+let beta2: expr = challenge(0, 4);
+
+let unpack_permutation_constraint: Constr -> (expr, expr[], expr, expr[]) = |permutation_constraint| match permutation_constraint {
+    Constr::Permutation(Option::Some(lhs_selector), lhs, Option::Some(rhs_selector), rhs) => (lhs_selector, lhs, rhs_selector, rhs),
```

Comment:
This is where I would have to handle the case were the selector is `None`. As proposed in the other comment, I think we should instead change the type of the selector.

Reply:
What's the problem with
```suggestion
    Constr::Permutation(lhs_selector, lhs, rhs_selector, rhs) => (unwrap_or_else(lhs_selector, || 1), lhs, unwrap_or_else(rhs_selector, || 1), rhs),
```
?